Module/sage/1.0

Kdreval opened the pull request:
# Pull Request Checklists

**Important:** When opening a pull request, keep only the applicable checklist and delete all other sections.

## Checklist for New Module

### Required

- [x] I used the cookiecutter template and updated the placeholder rules.

- [X] The snakemake rules follow the [design guidelines](https://lcr-modules.readthedocs.io/en/latest/for_developers.html#module-rules). 

  - [X] All references to the `rules` object (_e.g._ for input files) are wrapped with `str()`.

- [X] Every rule in the module is either listed under `localrules` or has the `threads` and `resources` directives.

- [X] Input and output files are being symlinked into the `CFG["inputs"]` and `CFG["outputs"]` subdirectories, respectively.

- [X] I updated the final target rule (`*_all`) to include every output rule.

- [X] I explained important module design decisions in `CHANGELOG.md`.

- [X] I tested the module on real data for all supported `seq_type` values and `genome_build`s.

- [X] I updated the `default.yaml` configuration file to provide default values for each rule in the module snakefile.

- [X] I did not set any global wildcard constraints. Any/all wildcard constraints are set on a per-rule basis. 

- [X] I ensured that all symbolic links are relative and self-contained (_i.e._ do not point outside of the repository). This includes environment files.

- [X] I replaced every value that should (or might need to) be updated in the default configuration file with `__UPDATE__`.

- [X] I recursively searched for all comments containing `TODO` to ensure none were left. For example:

  ```bash
  grep -r TODO modules/<module_name>/1.0
  ```

### If applicable

- [X] I added more granular output subdirectories.

- [X] I added rules to the `reference_files` workflow to generate any new reference files.

- [ ] I added subdirectories with large intermediate files to the list of `scratch_subdirectories` in the `default.yaml` configuration file.

- [ ] I updated the list of available wildcards for the input files in the `default.yaml` configuration file.

rdmorin commented on modules/sage/1.0/sage.smk:
```diff
@@ -0,0 +1,255 @@
+        -panel_bed {input.panel_bed}
+        -high_confidence_bed {input.high_conf_bed}
+        -out {output.vcf}
+        >>  {log.stdout} 2> {log.stderr}
```
It shouldn't matter here because stderr was only used once but I try to habitually use >> for all my redirects to log files.

Kdreval replied:
I think it is a safer way, I might miss something, or someone else can be modifying this snakefile later. So I changed in to `>>` throughout. I think it's a safe practice and I'll add it to other modules too!

rdmorin commented on modules/sage/1.0/sage.smk:
```diff
@@ -0,0 +1,255 @@
+        -threads {threads}
+        {params.opts}
+        -chr {params.chromosomes}
+        -reference {wildcards.normal_id}
```
I was puzzled by this at first. Reference is the name for the normal!

Kdreval replied:
Yes it is confusing to me as well! Why not name it `normal`, if it is fact not reference, but the normal sample.

rdmorin commented on modules/sage/1.0/sage.smk:
```diff
@@ -0,0 +1,255 @@
+        op.as_one_line("""
+        bcftools view -f ".,PASS" {input.vcf} -Ov
+          | 
+        bcftools sort --max-mem {resources.mem_mb}M -Oz -o {output.vcf_passed}
```
To be safe, should we use 0.8*mem_mb here as you did in the earlier example?

Kdreval replied:
good suggestion, I have added this to the rule

rdmorin commented on modules/sage/1.0/sage.smk:
```diff
@@ -0,0 +1,255 @@
+        **CFG["resources"]["filter"]
+    shell:
+        op.as_one_line("""
+        bcftools view -v indels {input.vcf_passed} -Oz -o {output.indels} > {log.stdout} 2> {log.stderr}
```
I had no idea you could use bcftools view this way. Nice!

rdmorin commented on modules/sage/1.0/sage.smk:
```diff
@@ -0,0 +1,255 @@
+          &&
+        bcftools view -v snps {input.vcf_passed} -Oz -o {output.snvs} >> {log.stdout} 2>> {log.stderr}
+          &&
+        tabix -p vcf {output.snvs} >> {log.stdout} 2>> {log.stderr}
```
Notably, you can also use "bcftools index"

rdmorin commented on modules/sage/CHANGELOG.md:
```diff
@@ -0,0 +1,16 @@
+- Reference files are symlinked to `00-inputs` from references directory to ensure all required indexes and dictionaries are created if starting on fresh directory.
+- SAGE fails to write to .gz files directly, therefore outputs are marked as `temp()` to ensure they are deleted after bgzip.
+- `main_chromosomes.txt` from reference_files workflow is used to restrict SAGE only to standard chromosomes. Chromosomes listed in this file are returned as list to be included with SAGE `-chr` flag.
+- module outputs separate VCF files for indels, snvs, and combined (indels+snsvs+mnvs). This can be modified in later releases.
```
what's a mnv?

Kdreval replied:
That's how they call multi-nucleotide variants, or clusters of 2 or more nearby variants.

rdmorin commented on modules/sage/1.0/sage.smk:
```diff
@@ -0,0 +1,255 @@
+          &&
+        tabix -p vcf {output.indels} >> {log.stdout} 2>> {log.stderr}
+          &&
+        bcftools view -v snps {input.vcf_passed} -Oz -o {output.snvs} >> {log.stdout} 2>> {log.stderr}
```
Does the sum of the variants in the two outputs add up to the combined or do we lose some here?

Kdreval replied:
It does, but not always since I do not subset vcf file for mnvs. In other words, if there are no mnvs, the sum adds up to the total number of variants. If there are mnvs, then sum of the 2 outputs+number of mnvs adds up to the number in `combined` vcf.

rdmorin replied:
I see. So we lose mnvs if we use the individual split files. That's not ideal but I expect these are pretty rare, right?

rdmorin commented on modules/sage/1.0/config/default.yaml:
```diff
@@ -0,0 +1,43 @@
+        pairing_config:
+            genome:
+                run_paired_tumours: True
+                run_unpaired_tumours_with: "unmatched_normal"
```
I agree with this implementation but I am wondering if we should also support running without a normal if that option is available. This could also be added to a later version.

rdmorin reviewed (changes requested):
Very clean first version. A few minor suggestions and comments. Ideally we should make version 1.0 support crams if we can.

rdmorin commented on modules/sage/1.0/sage.smk:
```diff
@@ -0,0 +1,255 @@
+        indels_tbi = CFG["dirs"]["outputs"] + "indels/{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}.sage.indels.vcf.gz.tbi",
+        snvs = CFG["dirs"]["outputs"] + "snvs/{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}.sage.snvs.vcf.gz",
+        snvs_tbi = CFG["dirs"]["outputs"] + "snvs/{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}.sage.snvs.vcf.gz.tbi"
+    run:
```
I urge you to consider changing this to use Laura's new "in_module" feature of relative_symlink as I did in the Battenberg module. Don't change it for the inputs but do it for all outputs and any intermediate steps that make symlinks.

Kdreval replied:
I have included this option, and as well included the check for oncopipe version.

Kdreval commented on modules/sage/1.0/sage.smk:
```diff
@@ -0,0 +1,255 @@
+        bam = CFG["inputs"]["sample_bam"]
+    output:
+        bam = CFG["dirs"]["inputs"] + "bam/{seq_type}--{genome_build}/{sample_id}.bam",
+        bai = CFG["dirs"]["inputs"] + "bam/{seq_type}--{genome_build}/{sample_id}.bam.bai"
```
I tested it and it works as-is with cram files, the only downside is increased running time from ~2.5h to ~ overnight.

Kdreval commented on modules/sage/1.0/sage.smk:
```diff
@@ -0,0 +1,255 @@
+    params:
+        opts = CFG["options"]["sage_run"],
+        chromosomes = get_chromosomes,
+        assembly = lambda w: "hg38" if "38" in str({w.genome_build}) else "hg19",
```
I am not sure why it needs this, but I found it is common with some other hmftools. It is a required argument, so I came up with this way of parsing it.

Kdreval commented on modules/sage/1.0/sage.smk:
```diff
@@ -0,0 +1,255 @@
+    params:
+        url = "www.bcgsc.ca/downloads/morinlab/hmftools-references/sage",
+        build = lambda w: "hg38" if "38" in str({w.genome_build}) else "hg19"
+    shell: 
```
It is not hard-coded to have hg in naming. We can supply a reference file with any name, and I opted for this way of naming the files to be consistent with the required argument `assembly`. I can create files with the actual `genome_build`, as in the wildcard, but I thought in this way I will need to create a reference file (or at least a symlink) for each genome build present in lcr-modules, and that might increase the number of files.

Kdreval commented:
> Very clean first version. A few minor suggestions and comments. Ideally we should make version 1.0 support crams if we can.

Thank you! I have addressed the comments and made some minor modifications. I also can confirm it works with cram files I tested. The updated version is ready for review.

Kdreval commented:
After I converted the vcfs to maf, I see it does not populate t_depth, t_ref_coun, t_alt_count. I will work on it to make them compatible.

Kdreval commented:
> After I converted the vcfs to maf, I see it does not populate t_depth, t_ref_coun, t_alt_count. I will work on it to make them compatible.

It was more related to vcf2maf config than to SAGE itself. I fixed it and it is populating reads now. No changes are required to the SAGE module.

rdmorin commented on modules/sage/1.0/config/default.yaml:
```diff
@@ -0,0 +1,43 @@
+        resources:
+            sage_run:
+                mem_mb: 60000
+                disk_mb: 20000 # Will ensure at least this much space is left
```
reserving/expecting 20G per sample seems a bit high, no?

Kdreval replied:
the largest file from the output is 3.5 Gb, but because at some point both compressed and decompressed files may exist (right before the _sage_run is completed), I gave it more space to accommodate both of these files. But I might decrease that further down if needed.

rdmorin commented:
Ideally this would be incorporated into the demo but that can wait since we really need this tool in production.

Kdreval commented:
> Ideally this would be incorporated into the demo but that can wait since we really need this tool in production.

I did not add it to the demo since the demo currently is being modified. I thought to include it in the demo with the demo development.

rdmorin commented:
Did you need this re-reviewed?

Kdreval commented:
I have included it in the demo and decreased the disk_mb requirement. I have also tested it outside of the GSC network and found that main_chromosomes need to be added to the rule. These are all modifications. If you approve it, it will be ready to go, so yes, please.
